### DISCUSSION TITLE
Is there a way to use airflow provider hooks within PythonVirtualenvOperator callable?

### DISCUSSION THREAD

**Discussion by ScriptPup** [apache/airflow#29290] [Q&A]
We've recently started using airflow and are trying to keep our environment clean. We have setup some providers and setup connections in order to interface with remote databases. This is working perfectly within "normal" operators. 

Now I have a use-case where a single DAG needs a package that's not installed globally. We do not want to install the package globally for a single DAG which isn't even going to be maintained for very long. My thought was the PythonVirtualenvOperator should be perfect. We can install the package within the task itself and we're good to go.

The problem we're running into is that it doesn't seem to be possible to use the provider hooks within the PythonVirtualenvOperator.

I've tried like this:

```python
...

## Python Callable
def myFunction():
    from my_provider.hooks.my_hook import myHook     
    db = myHook().get_conn()
    db.query("SELECT doSomething();")
    
# Python Operator
opr = PythonVirtualenvOperator(
    task_id = 'get_pip_libs'
    ,dag=dag
    ,python_callable=myFunction
    ,requirements=["arrow"]
)
opr.doc_md = """Check installed packages"""
```

The above fails with: `ModuleNotFoundError: No module named 'my_provider'`

I've also tried passing the hook through the ops (I figured it wouldn't work since I think options are pickled, but was worth a shot)
```python

...

## Python Callable
def myFunction(myHook):
    from my_provider.hooks.my_hook import myHook     
    db = myHook().get_conn()
    db.query("SELECT doSomething();")
    
# Python Operator
opr = PythonVirtualenvOperator(
    task_id = 'mytask'
    ,dag=dag
    ,python_callable=myFunction
    ,requirements=["arrow"]
    ,op_args=[myHook]
)
opr.doc_md = """Check installed packages"""
```

The above also fails with: `ModuleNotFoundError: No module named 'my_provider'`

I also tried to just reference the hooks in the callable directly...

```python
...

from my_provider.hooks.my_hook import myHook   

## Python Callable
def myFunction():
    db = myHook().get_conn()
    db.query("SELECT doSomething();")
    
# Python Operator
opr = PythonVirtualenvOperator(
    task_id = 'mytask'
    ,dag=dag
    ,python_callable=myFunction
    ,requirements=["arrow"]
)
opr.doc_md = """Check installed packages"""
```

The above also fails with: `NameError: name 'myHook' is not defined`

I've also tested each option with `system_site_packages = True` property set.

Is there any way to use our hooks directly within PythonVirtualenvOperator?

**Comment by potiuk** [comment]
You need to package them in a package and add it as PythonVirtualenvOperator requirements so that it is installed when virtual env is installed.
